Add 3dsx sending (in addition to receiving), ripped from 3dslink

LiquidFenrir opened the pull request:
Hello! This finally works perfectly. After our talk on irc, I changed it to use a light event and the swkbd, and it works like a charm.  
Unfortunately, couldn't add the ip autodetect, because SO_BROADCAST ([used here](https://github.com/devkitPro/3dslink/blob/master/host/src/main.c#L101)) doesn't exist, or at least that's what the compilation error said. I'm a noob with sockets, so couldn't fix that.  
Let me know if there's anything you don't like in there and if it can be fixed.  
I think this feature would be nice to have so people can share homebrew between their consoles (there's also uds that would allow that without needing wifi, but it's really unreliable).

LiquidFenrir committed:
works/10
TODO:
- scanning for consoles
- swkbd for ip entering if above fails
- fix strings
- add error reporting

LiquidFenrir committed:
only allow sending 3dsx files

LiquidFenrir committed:
style fix for if statements, and close everything when erroring out of netsenderTask

LiquidFenrir committed:
this should work, but it don't

LiquidFenrir committed:
fix text to indicate sending. needs translations

LiquidFenrir committed:
inputting the ip now uses swkbd, and works.
text for invalid IP needs translations

LiquidFenrir committed:
fix spaces -> tabs
copy paste error

LiquidFenrir committed:
code consistency

LiquidFenrir committed:
add autodetection for other consoles

LiquidFenrir committed:
replace fprintf to stderr with proper netsenderError

LiquidFenrir committed:
less duplication for text, and showing a screen when searching

fincs commented on source/language.c:
```diff
@@ -795,4 +795,104 @@ const char* const g_strings[StrId_Max][16] =
+		STR_TW(""),
+	},
+	
+	[StrId_NetSenderActive] =
```
This is a duplicate of StrId_NetLoaderActive.

LiquidFenrir replied:
not exactly a duplicate because the format part for the ip and port was removed. I believe they should stay separate, as showing the ip is useless for the sender

fincs commented on source/ui/menu.c:
```diff
@@ -104,6 +105,16 @@ void menuUpdate(void)
+		menuEntry_s* me;
+		for (i = 0, me = menu->firstEntry; i != menu->curEntry; i ++, me = me->next);
+		if(me->type == ENTRY_TYPE_FILE)
+		{
```
Redundant braces

fincs commented on source/ui/menu.c:
```diff
@@ -104,6 +105,16 @@ void menuUpdate(void)
+		int i;
+		menuEntry_s* me;
+		for (i = 0, me = menu->firstEntry; i != menu->curEntry; i ++, me = me->next);
+		if(me->type == ENTRY_TYPE_FILE)
```
Space after `if`

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+
+static void netsenderError(const char* func, int err);
+
+static bool netsenderInit(void)
```
I wonder if netsenderInit/netsenderDeactivate could be conflated with their netloader* counterparts since they have the same code.

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+---------------------------------------------------------------------------------*/
+static int timeval_subtract (struct timeval *result, struct timeval *x, struct timeval *y)
+{
+//---------------------------------------------------------------------------------
```
Remove this line

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+	return x->tv_sec < tmp.tv_sec;
+}
+
+//---------------------------------------------------------------------------------
```
Remove these `//----` comment lines

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+	tmp.tv_sec = y->tv_sec;
+	tmp.tv_usec = y->tv_usec;
+
+	/* Perform the carry for the later subtraction by updating y. */
```
Convert to `//` style comments

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+		tmp.tv_sec -= nsec;
+	}
+
+	/*	Compute the time remaining to wait.
```
Same here (and put in a single line instead of two)

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+	result->tv_sec = x->tv_sec - tmp.tv_sec;
+	result->tv_usec = x->tv_usec - tmp.tv_usec;
+
+	/* Return 1 if result is negative. */
```
Same

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+	}
+}
+
+//---------------------------------------------------------------------------------
```
Same thing, etc

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+		return remote.sin_addr;
+	}
+
+	memset(&s, '\0', sizeof(struct sockaddr_in));
```
Use `0` instead of `'\0'`

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+	s.sin_port = htons(17491);
+	s.sin_addr.s_addr = INADDR_BROADCAST;
+
+	memset(&rs, '\0', sizeof(struct sockaddr_in));
```
Same

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+	gettimeofday(&wanted, NULL);
+
+	int timeout = retries, len;
+	while(timeout)
```
Space after `while`

fincs replied:
Currently this while loop is a busy loop that will burn the CPU. Could you please add a small `svcSleepThread` at the end of the loop body?

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+		if (len != -1)
+		{
+			if (strncmp("boot3ds",recvbuf,strlen("boot3ds")) == 0)
+			{
```
Remove useless braces

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+
+int sendData(int sock, int sendsize, char *buffer)
+{
+	while(sendsize)
```
Space after `while`

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+	}
+
+	struct sockaddr_in s;
+	memset(&s, '\0', sizeof(struct sockaddr_in));
```
Same thing

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+
+	if (response != 0)
+	{
+		switch(response) {
```
Space after while, brace on new line  (heh, that rhymed)

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+
+	size_t totalsent = 0, blocks = 0;
+
+	do {
```
Brace on new line

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+	do {
+		strm.avail_in = fread(in, 1, ZLIB_CHUNK, fh);
+		filetotal += strm.avail_in;
+		if (ferror(fh)) {
```
Brace on new line

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+		filetotal += strm.avail_in;
+		if (ferror(fh)) {
+			netsenderError("ferror", 0);
+			(void)deflateEnd(&strm);
```
No need to cast to void

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+		strm.next_in = in;
+		/* run deflate() on input until output buffer not full, finish
+		   compression if all of source has been read in */
+		do {
```
Brace on new line

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+		}
+		flush = feof(fh) ? Z_FINISH : Z_NO_FLUSH;
+		strm.next_in = in;
+		/* run deflate() on input until output buffer not full, finish
```
Use `//`

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+				if (sendData(datafd, have, (char*)out))
+				{
+					netsenderError("have", errno);
+					(void)deflateEnd(&strm);
```
Same thing

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+				blocks++;
+			}
+		} while (strm.avail_out == 0);
+		/* done when last data in file processed */
```
Same thing

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+		} while (strm.avail_out == 0);
+		/* done when last data in file processed */
+	} while (flush != Z_FINISH);
+	(void)deflateEnd(&strm);
```
Same thing

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+		return;
+	}
+	
+	close(datafd);
```
This code is repeated over and over in the error cases. Instead of that, could you add a label here and goto from the other places? That's the customary way to do it in C.

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+
+	dsaddr = find3DS(10);
+	doneSearching = true;
+	if(dsaddr.s_addr == INADDR_NONE)
```
Space after if

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+	doneSearching = true;
+	if(dsaddr.s_addr == INADDR_NONE)
+	{
+		LightEvent_Init(&event, RESET_ONESHOT);
```
LightEvent_Init shouldn't be done over and over. Please do it in netsenderInit.

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+
+	netsenderDeactivate();
+	uiExitState();
+	return;
```
Useless return

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+		return inet_addr(ip) != 0;
+	}
+	else
+	{
```
Useless else clause and braces

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+		*ppMessage = textGetString(StrId_NetSenderInvalidIp);
+		return SWKBD_CALLBACK_CONTINUE;
+	}
+	else
```
Useless else clause and braces

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+
+static SwkbdCallbackResult callback(void *user, const char **ppMessage, const char *text, size_t textlen)
+{
+	(void)user;
```
Useless statement

LiquidFenrir replied:
gets rid of the unused parameter warning, so not really useless

fincs replied:
There's no unused parameter warning. I don't think we're using `-Wextra`. If we are, please remove it.

LiquidFenrir replied:
you are right, you only use -Wall. my bad, habit from using both. sorry

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+	if (wantExit || datafd >= 0) return;
+
+	if (hidKeysDown() & KEY_B)
+	{
```
Useless braces

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+		wantExit = true;
+	}
+
+	if(doneSearching && dsaddr.s_addr == INADDR_NONE)
```
Space after if

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+		swkbdSetFilterCallback(&swkbd, callback, NULL);
+		SwkbdButton button = swkbdInputText(&swkbd, receiverIp, RECEIVER_IP_LENGTH+1);
+		if (button == SWKBD_BUTTON_LEFT) // Cancel
+		{
```
Useless braces

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+
+	textDraw(8.0f, 60.0f+25.0f+8.0f, 0.5f, 0.5f, false, text);
+
+	if (datafd >= 0 && filelen)
```
Most of this code is duplicated from netloader. Could there be a way it could be shared?

fincs commented on source/language.c:
```diff
@@ -795,4 +795,104 @@ const char* const g_strings[StrId_Max][16] =
+
+	[StrId_NetSender] =
+	{
+		STR_EN("3dslink NetSender"),
```
Minor terminology nitpick: I think I would call this functionality a "3dslink server" instead of a "netsender". I don't think I've ever seen anything called a "netsender".

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,607 @@
+
+static void netsenderDeactivate(void)
+{
+	socExit();
```
There's a memory leak here. SOC_buffer from the above function is never freed.

fincs reviewed (changes requested):
Overall, thank you for your contribution. However before it can be merged, there are several things that need to be fixed.

LiquidFenrir committed:
fix code style for menu.c

LiquidFenrir committed:
netsender -> 3dslink server for displayed text

LiquidFenrir committed:
start putting common network code in its own file

LiquidFenrir committed:
memset '\0' -> 0

LiquidFenrir committed:
change comment style for find3DS related functions

LiquidFenrir committed:
remove useless "separation" comments

LiquidFenrir committed:
make networkError more general

LiquidFenrir committed:
remove now unused #include <malloc.h> and put it where needed

LiquidFenrir committed:
attempt at sharing drawBot

LiquidFenrir committed:
code style fixes and getting rid of useless braces

LiquidFenrir committed:
prevent double freeing (could crash)

LiquidFenrir committed:
add a 1 millisecond sleep when trying to detect consoles

LiquidFenrir committed:
add spaces between args

LiquidFenrir committed:
space after while and switch

LiquidFenrir committed:
-Wextra isn't there, you dont need to void unused args

LiquidFenrir committed:
actually cancel if user presses B while searching for consoles

LiquidFenrir committed:
more memset

LiquidFenrir committed:
brace on newline

LiquidFenrir committed:
no need to cast to void

LiquidFenrir committed:
use goto for erroring out of send3DSXFile

LiquidFenrir committed:
more braces on more newlines, and more unneeded voiding

fincs commented on source/network.c:
```diff
@@ -0,0 +1,84 @@
+void networkDeactivate(void)
+{
+	socExit();
+	if(SOC_buffer)
```
Space after if

fincs commented on source/network.c:
```diff
@@ -0,0 +1,84 @@
+
+	char buf[256];
+	const char* text = other;
+	if(text == NULL)
```
Space after if

fincs commented on source/network.h:
```diff
@@ -0,0 +1,9 @@
+bool networkInit(void);
+void networkDeactivate(void);
+void networkError(void (* update)(void), StrId titleStrId, const char* func, int err);
+void networkDrawBot(StrId titleStrId, const char* other, bool transferring, size_t filelen, size_t filetotal);
```
Should have a newline at eof

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,537 @@
+
+	if (response != 0)
+	{
+		switch (response) {
```
Newline before brace

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,537 @@
+		}
+		flush = feof(fh) ? Z_FINISH : Z_NO_FLUSH;
+		strm.next_in = in;
+		/* run deflate() on input until output buffer not full, finish
```
Missed `//`

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,537 @@
+		{
+			strm.avail_out = ZLIB_CHUNK;
+			strm.next_out = out;
+			ret = deflate(&strm, flush);    /* no bad return value */
```
Use `//` instead of `/* */`

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,537 @@
+	if (sendData(datafd,cmdlen+4,cmdbuf))
+		netsenderError("cmdbuf", errno);
+	
+	send3DSXFileExit:
```
Labels aren't indented, should be on column 0. Also, I'd call this label `_cleanup:`

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,537 @@
+				blocks++;
+			}
+		}
+		while (strm.avail_out == 0);
```
My code style does do..while loops like this:
```c
do
{
    //...
} while (condition);
```

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,537 @@
+		while (strm.avail_out == 0);
+		/* done when last data in file processed */
+	}
+	while (flush != Z_FINISH);
```
Same thing

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,537 @@
+	fclose(inf);
+	free(filename);
+
+	netsenderDeactivate();
```
Might want to simplify error handling with goto here too

fincs commented on source/ui/netsender.c:
```diff
@@ -165,6 +165,8 @@ static struct in_addr find3DS(int retries)
 			}
 		}
+
+		svcSleepThread(1e6);
```
This is actually a `double` floating point constant. I'd advise you to write out the zeroes `1000000`. Actually I'd make this 10ms and not 1ms: `10000000`

LiquidFenrir committed:
newline at eof for network.h

LiquidFenrir committed:
space after if, again

LiquidFenrir committed:
fix style for comments and do while

LiquidFenrir committed:
label indent fix, change name, and add for netsenderTask

LiquidFenrir committed:
brace on newline for switch

LiquidFenrir committed:
fix label name in netsenderTask, they're function local

LiquidFenrir committed:
NetLoader -> Network for strings no longer NetLoader-specific

fincs commented on source/ui/netloader.c:
```diff
@@ -410,41 +392,14 @@ void netloaderExit(void)
+		text = buf;
 	}
+
+	networkDrawBot(StrId_NetSender, text, (datafd >= 0 && filelen), filelen, filetotal);
```
Should be `StrId_NetLoader` :p

fincs commented on source/ui/netsender.c:
```diff
@@ -0,0 +1,528 @@
+		{
+			if (sendto(broadcastSock, mess, strlen(mess), 0, (struct sockaddr *)&s, sizeof(s)) < 0)
+			{
+				netsenderError("sendto", errno);
```
I tested, and apparently this error breaks the UI logic, because uiExitState ends up being called twice (one indirectly here through netsenderError, and the other in netsenderTask). Could you look into it? (I think you should somehow add a guard so that netsenderTask doesn't call uiExitState if netsenderError has already been used).

LiquidFenrir committed:
copy paste error, NetSender -> NetLoader

LiquidFenrir committed:
check if netsenderError already happened before using uiExitState
also, directly exit out of netsenderTask instead of waiting on an event that will never signal if you error out of find3DS
